### DISCUSSION TITLE
Rook RGW Perf Test

### DISCUSSION THREAD

**Discussion by hawk87** [rook/rook#10847] [General]
Hi everyone!

I'm evaluating the performance of Rook Object Storage with Warp (https://github.com/minio/warp)
I'm using CephFS

The CephObjectStore CRD is deployed w/ object-test.yaml first, then object-ec.yaml (https://github.com/rook/rook/blob/release-1.9/deploy/examples/object-test.yaml and https://github.com/rook/rook/blob/release-1.9/deploy/examples/object-ec.yaml) and didn't notice substantial differences during the benchmark.

It seems that Rook has very poor performance compared to other S3 solutions.

```
[rook@rook-ceph-tools-545fd87745-8b9pw tmp]$ ./warp mixed --host=$AWS_HOST --access-key=$AWS_ACCESS_KEY_ID --secret-key=$AWS_SECRET_ACCESS_KEY --autoterm --duration=5m --analyze.v --objects 500
warp: Benchmark data written to "warp-mixed-2022-08-09[135134]-swVv.csv.zst"                                                                                                                                     
Mixed operations.
Operation: DELETE - total: 498, 9.8%, Concurrency: 20, Ran 4m50s, starting 2022-08-09 13:51:43.143 +0000 UTC
 * Throughput: 1.74 obj/s

Requests considered: 499:
 * Avg: 1.839s, 50%: 1.665s, 90%: 2.899s, 99%: 6.61s, Fastest: 463ms, Slowest: 9.818s

Operation: GET - total: 2269, 44.8%, Concurrency: 20, Ran 4m52s, starting 2022-08-09 13:51:41.247 +0000 UTC
 * Throughput: 78.01 MiB/s, 7.80 obj/s

Requests considered: 2270:
 * Avg: 1.238s, 50%: 1.125s, 90%: 1.982s, 99%: 3.748s, Fastest: 51ms, Slowest: 9.121s
 * TTFB: Avg: 529ms, Best: 15ms, 25th: 234ms, Median: 468ms, 75th: 704ms, 90th: 999ms, 99th: 2.128s, Worst: 3.489s
 * First Access: Avg: 1.297s, 50%: 1.195s, 90%: 2.017s, 99%: 4.32s, Fastest: 60ms, Slowest: 9.121s
 * First Access TTFB: Avg: 549ms, Best: 18ms, 25th: 224ms, Median: 481ms, 75th: 714ms, 90th: 1.096s, 99th: 2.498s, Worst: 3.489s
 * Last Access: Avg: 1.204s, 50%: 1.109s, 90%: 1.9s, 99%: 3.628s, Fastest: 122ms, Slowest: 4.795s
 * Last Access TTFB: Avg: 518ms, Best: 15ms, 25th: 224ms, Median: 432ms, 75th: 696ms, 90th: 995ms, 99th: 2.297s, Worst: 3.28s

Operation: PUT - total: 747, 14.7%, Concurrency: 20, Ran 4m52s, starting 2022-08-09 13:51:41.301 +0000 UTC
 * Throughput: 25.97 MiB/s, 2.60 obj/s

Requests considered: 748:
 * Avg: 2.42s, 50%: 2.284s, 90%: 3.292s, 99%: 5.796s, Fastest: 1.023s, Slowest: 8.007s

Operation: STAT - total: 1529, 30.2%, Concurrency: 20, Ran 4m52s, starting 2022-08-09 13:51:41.24 +0000 UTC
 * Throughput: 5.23 obj/s

Requests considered: 1530:
 * Avg: 158ms, 50%: 35ms, 90%: 486ms, 99%: 1.412s, Fastest: 1ms, Slowest: 2.874s
 * First Access: Avg: 151ms, 50%: 29ms, 90%: 481ms, 99%: 1.319s, Fastest: 1ms, Slowest: 2.098s
 * Last Access: Avg: 139ms, 50%: 41ms, 90%: 410ms, 99%: 1.103s, Fastest: 1ms, Slowest: 1.788s

Cluster Total: 103.71 MiB/s, 17.32 obj/s over 4m53s.
```
Am I doing something wrong or these are the limits achievable by Rook?

Thank you in advance!

**Comment by travisn** [comment]
> I'm using CephFS

What do you mean you're using CephFS? You're only testing the object store, right?

> The CephObjectStore CRD is deployed w/ object-test.yaml first, then object-ec.yaml 

object-test.yaml only uses replica 1 (no data redundancy), while object-ec.yaml uses replica 3 for the metadata and EC 2+1 for the data. All other things being equal, object-test should perform better since there is some overhead for the EC. But you're not seeing any difference?

How many RGW pods do you have running? To get more parallelism, try running more rgw daemons. If you just have one daemon, everything will be gated on that one daemon.

What media is backing the OSDs? hdds or ssds? There are several such factors such as network in the infrastructure that could be limiting the throughput as well.

**Reply by hawk87** [reply]
Hi @travisn , thank you for your reply.
I found out that I was running my performance tests on VM snapshot (it degrades the performances by ~70%).
I run the tests again (also with more rgw daemons) and it seems that it's working as expected.

Thank you!
